sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala

```diff
@@ -274,6 +278,15 @@ private[orc] class OrcOutputWriter(
 
   override def close(): Unit = {
     if (recordWriterInstantiated) {
+      // Hive 1.2.1 ORC initializes its private `writer` field at the first write.
+      try {
+        val writerField = recordWriter.getClass.getDeclaredField("writer")
+        writerField.setAccessible(true)
+        val writer = writerField.get(recordWriter).asInstanceOf[Writer]
+        writer.addUserMetadata(SPARK_VERSION_METADATA_KEY, UTF_8.encode(SPARK_VERSION_SHORT))
+      } catch {
+        case NonFatal(e) => log.warn(e.toString, e)
+      }
```

Comment:
The same comment here.

Reply:
BTW, as you expected, we cannot use a single function for this. The `Writer` are not the same.

Reply:
For this case, I'll refactor out all the new code (line 281 ~ 289).